### PR TITLE
Port of AJAX stateless artifacts from wicketstuff stateless module

### PR DESCRIPTION
**Pull request by bitstorm**
Port of AJAX stateless artifacts from wicketstuff stateless module

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/ajax/StatelessOnChangeAjaxBehavior.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ * This behavior uses best available method to track changes on different types of form components.
+ * To accomplish this for text input form components it uses a custom event, named 'inputchange',
+ * that is handled in the best way for the specific browser. For other form component types the
+ * 'change' event is used.
+ * </p>
+ * 
+ * @author Andrea Del Bene
+ *
+ */
+public class StatelessOnChangeAjaxBehavior extends StatelessAjaxFormComponentUpdatingBehavior
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/ajax/StatelessOnChangeAjaxBehavior.java]: Except the parent class this class is an exact copy of `OnChangeAjaxBehavior`.
Can't we extract the common code somewhere and reuse it ?

File: wicket-core/src/main/java/org/apache/wicket/ajax/form/StatelessAjaxFormComponentUpdatingBehavior.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        final Url url = Url.parse(super.getCallbackUrl().toString());
+        final PageParameters params = getPageParameters();
+
+        return UrlUtils.mergeParameters(url, params).toString();
+    }
+
+    protected PageParameters getPageParameters()
+    {
+    	return null;
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/ajax/form/StatelessAjaxFormComponentUpdatingBehavior.java]: Why not `getComponent().getPageParameters()` ?

File: wicket-core/src/main/java/org/apache/wicket/ajax/form/StatelessAjaxFormComponentUpdatingBehavior.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public CharSequence getCallbackUrl()
+    {
+        final Url url = Url.parse(super.getCallbackUrl().toString());
+        final PageParameters params = getPageParameters();
+
+        return UrlUtils.mergeParameters(url, params).toString();
+    }
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/ajax/form/StatelessAjaxFormComponentUpdatingBehavior.java]: `#onBind()` and `#getCallbackUrl()` are the same as in StatelessAjaxEventBehavior.
Maybe we could introduce an interface with default methods that is mixed into those classes.

File: wicket-core/src/main/java/org/apache/wicket/ajax/form/StatelessAjaxFormSubmitBehavior.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+/**
+ * Stateless version of {@link AjaxFormSubmitBehavior}.
+ */
+public class StatelessAjaxFormSubmitBehavior extends StatelessAjaxEventBehavior
+{
+
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = -4052276703516669553L;
```
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/ajax/form/StatelessAjaxFormSubmitBehavior.java]: Useless javadoc.

**Comment by martin-g** [comment]
Overall it is OK, but I don't like that there is so much duplicated code.

**Comment by bitstorm** [comment]
Hi @martin-g and thank you for your reviw,

as you noted the code contains some classes where code is repeated. I've tried to reduce code duplication but without using Java 8 is nearly impossible. As you mentioned above default methods might help a lot. I'm considering to post-pone this integration to Wicket 8 where I'm free to use Java 8.

**Comment by martin-g** [comment]
Oh. I didn't pay attention to the destination branch. It is 'wicket-7.x', I thought it is 'master'.

**Comment by bitstorm** [comment]
@martin-g  using inheritance instead of a separate stateless hierarchy reduced the code a lot :-)

File: wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java [hunk 5]
```diff
@@ -81,7 +82,27 @@
...
+		}
+	}
+	
+	@Override
+	public CharSequence getCallbackUrl()
+	{
+		if (getStatelessHint(getComponent()))
+		{
+		    final Url url = Url.parse(super.getCallbackUrl().toString());
+	        return UrlUtils.mergeParameters(url).toString();
```
Comment (svenmeier) [inline wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: #mergeParameters() does nothing when no parameters are passed - are the page parameters missing here?

Comment (bitstorm) [inline wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: yes you are right, I haven't finished yet the refactoring. We can simplify even more moving away getCallbackUrl from AbstractDefaultAjaxBehavior and using it with page parameters only in StatelessAjaxEventBehavior.

Comment (bitstorm) [inline wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: @svenmeier yes you are just right. We should pass page parameters here.

**Comment by svenmeier** [comment]
Actually I'd fine with adding the stateless-handling into AbstractDefaultAjaxBehavior as proposed here.

But I'd rather not add all these Stateless* classes - IMHO it's fine to let users override #isStateless() when they want to support stateless requests.

**Comment by bitstorm** [comment]
BTW I'm starting to wonder if we really need a custom AbstractDefaultAjaxBehavior#getCallbackUrl() to add page parameters to the callback url. When we generate an URL for stateless pages we end up using  AbstractBookmarkableMapper or MountedMapper and both uses page.getPageParameters() to build the URL (line 182 for MountedMapper and 476 for AbstractBookmarkableMapper )

**Comment by bitstorm** [comment]
@svenmeier  I also agree that the best way to proceed now would be to commit just the code to make stateless AJAX possible, i.e. to commit just the code involving the existing AJAX components/behaviors. 
If everybody agrees I will close this PR without merging anything and I will open an issue on JIRA to commit just the code we need.

**Comment by martin-g** [comment]
I agree!

**Pull request closed without merging**